draco/asp/generate.lp

```diff
@@ -1,19 +1,27 @@
 % ====== Generators ======
 
+% helpers to generate attributes based on how they are required.
+{ attribute(N,root,V) : domain(N,V) } = 1 :- root_required(N).
+{ attribute((N,A),E,V): domain((N,A),V) } = 1 :- entity(N,_,E), required((N,A)).
+0 { attribute((N,A),E,V): domain((N,A),V) } 1 :- entity(N,_,E), not_required((N,A)).
+
 % single view v0 on the root
 { entity(view,root,(v,0)) } = 1.
 
-% maximum number of additional marks for each view
-#const max_extra_marks = 1.
-mark_id(0..max_extra_marks).
-{ entity(mark,V,(m,M)) : entity(view,_,V), mark_id(M) }.
-:- not entity(mark,_,(m,ID)), entity(mark,_,(m,ID-1)), mark_id(ID), mark_id(ID-1).
+% maximum number of non-layered views.
+#const max_views = 1.
+{ entity(view,root,(v,V)) } :- V = 0..(max_views-1).
+:- entity(view,root,(v,ID)), not entity(view,root,(v,ID-1)), ID > 1.
+
+% maximum number of marks for each view
+#const max_marks = 2.
+{ entity(mark,(v,V),(m,M)) } :- entity(view,_,(v,V)), N = 0..(max_marks-1), M = max_marks * V + N.
+:- entity(mark,(v,V),(m,ID)), not entity(mark,(v,V),(m,ID-1)), ID > 1 + max_marks * V.
 
 % maximum number for total encoding channels
-#const max_extra_encs = 4.
-encoding_id(M,E) :- entity(mark,_,(m,M)), N = 0..max_extra_encs, E = ((max_extra_encs + 1) * M) + N.
-{ entity(encoding,(m,M),(e,E)) : entity(mark,_,(m,M)), encoding_id(M,E) }.
-:- not entity(encoding,_,(e,ID)), entity(encoding,_,(e,ID-1)),encoding_id(_,(e,ID)), encoding_id(_,(e,ID-1)).
+#const max_encs = 4.
+{ entity(encoding,(m,M),(e,E)) } :- entity(mark,_,(m,M)), N = 0..(max_encs-1), E = max_encs * M + N.
```

Comment:
What's the purpose of `E = max_encs * M + N`?

Reply:
It generates N encodings for each mark.
ex. If there are 2 marks: (m,0) and (m,1). N=4. 
Then we generate (e,0) to (e,3) for (m,0), and (e,4) to (e,7) for (m,1).
Although one problem is that since N is not enforced, there could be gaps in the ids, like (e,0) for (m,0), and (e,4),(e,5) for (m,1) such that (e,1..3) doesn't exist in the sequence. 
Is there a better way to avoid this?

Reply:
Hmm, interesting. I guess we could add the mark to the ID tuple so that we could have numbers that are consecutive and start from 0. So like (e,2,0) where 2 is the mark and 0 is the encoding id.

Reply:
ha. I asked about this two weeks ago. Since we are having multiple views, things could become like (v,0), (m,0,0), (e,0,0,0)

Reply:
I think this could work. The other option is instead to limit the number of new encodings overall (not per view). That could also work but may in the end work less well. Not sure.

Reply:
done